lib/session.js

```diff
@@ -0,0 +1,262 @@
+    }
+
+    const stateKey = options.stateKey || 'state';
+    const nonceKey = options.nonceKey || 'nonce';
+    const urlPrefix = options.urlPrefix || '';
+    const sessionStorageKey = options.sessionStorageKey || 'apiToken';
+    const sessionManager = options.sessionManager ||
+      new tools.SessionManager(
+        options.rta,
+        options.domain,
+        options.baseUrl
+      );
+    const basicCookieAttr = {
+      isHttpOnly: true
+    };
+    server.state(nonceKey, Object.assign({}, basicCookieAttr, { isSameSite: 'None', isSecure: true }));
+    server.state(stateKey, Object.assign({}, basicCookieAttr, { isSameSite: 'None', isSecure: true }));
+    server.state(`${nonceKey}_compat`, basicCookieAttr);
+    server.state(`${stateKey}_compat`, basicCookieAttr);
+
+    server.route({
+      method: 'GET',
+      path: `${urlPrefix}/login`,
+      options: {
+        auth: false
+      },
+      handler: (req, h) => {
+        const state = crypto.randomBytes(16).toString('hex');
+        const nonce = crypto.randomBytes(16).toString('hex');
+        const redirectTo = sessionManager.createAuthorizeUrl({
+          redirectUri: buildUrl([urlHelpers.getBaseUrl(req), urlPrefix, '/login/callback']),
+          scopes: options.scopes,
+          expiration: options.expiration,
+          nonce,
+          state
+        });
+
+        return h.redirect(redirectTo)
+          .state(nonceKey, nonce, { path: urlHelpers.getBasePath(req) })
+          .state(stateKey, state, { path: urlHelpers.getBasePath(req) })
+          .state(`${nonceKey}_compat`, nonce, { path: urlHelpers.getBasePath(req) })
+          .state(`${stateKey}_compat`, state, { path: urlHelpers.getBasePath(req) });
+      }
+    });
+
+    server.route({
+      method: 'POST',
+      path: `${urlPrefix}/login/callback`,
+      options: {
+        auth: false
+      },
+      handler: async (req, h) => {
+        let decoded;
+
+        const jwtVerifyAsync = promisify(jwt.verify);
+        const idToken = req.payload.id_token;
+
+        try {
+          decoded = jwt.decode(idToken, { complete: true });
+          const getKey = jwksRsa.hapiJwt2Key({
+            cache: true,
+            rateLimit: true,
+            jwksRequestsPerMinute: process.env.NODE_ENV === 'test' ? 10 : 2,
+            jwksUri: `${config('AUTH0_RTA')}/.well-known/jwks.json`
+          });
+
+          const getKeyAsync = promisify(getKey);
+          const key = await getKeyAsync(decoded);
+
+          if (!key) {
+            return Boom.unauthorized('Invalid token');
+          }
+
+          // this token is issued by the RTA
+          const verifyOptions = {
+            audience: config('PUBLIC_WT_URL'),
+            issuer: `${config('AUTH0_RTA')}/`,
+            algorithms: ['RS256']
+          };
+
+          // // throws on failure
+          await jwtVerifyAsync(idToken, key, verifyOptions);
```

Comment:
I think you can just pass your get key function into this method. Would allow for the decoded JWT to be only done once and is coupled to verification. From the `jsonwebtoken` documentation

```
var client = jwksClient({
  jwksUri: 'https://sandrino.auth0.com/.well-known/jwks.json'
});
function getKey(header, callback){
  client.getSigningKey(header.kid, function(err, key) {
    var signingKey = key.publicKey || key.rsaPublicKey;
    callback(null, signingKey);
  });
}

jwt.verify(token, getKey, options, function(err, decoded) {
  console.log(decoded.foo) // bar
});
````

Reply:
@gausnes 
```
          const getKeyAsync = promisify(getKey);

          // this token is issued by the RTA
          const verifyOptions = {
            audience: config('PUBLIC_WT_URL'),
            issuer: `${config('AUTH0_RTA')}/`,
            algorithms: ['RS256']
          };
          await jwtVerifyAsync(idToken, getKeyAsync, verifyOptions);
```

Reply:
@gausnes Just fixed this using what you mentioned. Dynamic keys I guess can't use promisify so using callbacks